src/distributions/weighted_index.rs

```diff
@@ -236,6 +239,60 @@ where X: SampleUniform + PartialOrd
     }
 }
 
+/// Bounds on a weight
+///
+/// See usage in [`WeightedIndex`].
+pub trait Weight: Clone {
+    /// Representation of 0
+    const ZERO: Self;
+
+    /// Checked addition
+    ///
+    /// -   `Result::Ok`: On success, `v` is added to `self`
+    /// -   `Result::Err`: Returns an error when `Self` cannot represent the
+    ///     result of `self + v` (i.e. overflow). The value of `self` should be
+    ///     discarded.
+    fn checked_add_assign(&mut self, v: &Self) -> Result<(), ()>;
```

Comment:
The only thing that concerns me slightly is that this is a fairly complex signature that, as you say, we can't easily change later.

Possibly the standard [`checked_add`](https://doc.rust-lang.org/stable/std/primitive.i32.html#method.checked_add) signature would be better, but this one is closer to what we actually need (considering we don't have a `Copy` bound).